source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,582 @@
+                 xMQTTPublishInfo->topicNameLength,
+                 xMQTTPublishInfo->pTopicName,
+                 xMQTTPublishInfo->payloadLength,
+                 xMQTTPublishInfo->pPayload ) );
+
+    if( pxContext->_internal.callbacks.xMethodCallback )
+    {
+        xMethodRequest.pvMessagePayload = xMQTTPublishInfo->pPayload;
+        xMethodRequest.ulPayloadLength = ( uint32_t ) xMQTTPublishInfo->payloadLength;
+        xMethodRequest.pucMethodName = az_span_ptr( xOutEmbeddedRequest.name );
+        xMethodRequest.usMethodNameLength = ( uint16_t ) az_span_size( xOutEmbeddedRequest.name );
+        xMethodRequest.pucRequestID = az_span_ptr( xOutEmbeddedRequest.request_id );
+        xMethodRequest.usRequestIDLength = ( uint16_t ) az_span_size( xOutEmbeddedRequest.request_id );
+        pxContext->_internal.callbacks.xMethodCallback( &xMethodRequest, pxContext->_internal.pvCallbackContext );
+    }
+
+    return eAzureIoTHubClientSuccess;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Check/Process messages for incoming device twin messages.
+ *
+ * */
+static uint32_t prvAzureIoTHubClientDeviceTwinProcess( AzureIoTHubClientReceiveContext_t * pxContext,
+                                                       AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                       void * pxPublishInfo )
+{
+    AzureIoTHubClientTwinResponse_t xTwinResponse;
+    AzureIoTMQTTPublishInfo_t * xMQTTPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pxPublishInfo;
+    az_iot_hub_client_twin_response xOutRequest;
+    az_span xTopicSpan = az_span_create( ( uint8_t * ) xMQTTPublishInfo->pTopicName, xMQTTPublishInfo->topicNameLength );
+    uint32_t xRequestID = 0;
+
+    /* Failed means no topic match */
+    if( az_result_failed( az_iot_hub_client_twin_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                       xTopicSpan, &xOutRequest ) ) )
+    {
+        return eAzureIoTHubClientTopicNoMatch;
+    }
+
+    /* Process incoming Publish. */
+    AZLogInfo( ( "Twin Incoming QoS : %d.", xMQTTPublishInfo->qos ) );
+
+    /* Verify the received publish is for the feature we have subscribed to. */
+    AZLogInfo( ( "Twin topic: %.*s. with payload : %.*s.",
+                 xMQTTPublishInfo->topicNameLength,
+                 xMQTTPublishInfo->pTopicName,
+                 xMQTTPublishInfo->payloadLength,
+                 xMQTTPublishInfo->pPayload ) );
+
+    if( pxContext->_internal.callbacks.xTwinCallback )
+    {
+        if( az_span_size( xOutRequest.xRequestID ) == 0 )
+        {
+            xTwinResponse.messageType = eAzureIoTHubTwinDesiredPropertyMessage;
+        }
+        else
+        {
+            if( az_result_failed( az_span_atou32( xOutRequest.xRequestID, &xRequestID ) ) )
+            {
+                return eAzureIoTHubClientFailed;
+            }
```

Comment:
how about arranging it like:

static uint32_t prvAzureIoTHubClientDeviceTwinProcess( AzureIoTHubClientReceiveContext_t * pxContext,
                                                       AzureIoTHubClient_t * pxAzureIoTHubClient,
                                                       void * pxPublishInfo )
{
    AzureIoTHubClientTwinResponse_t xMessage;
    AzureIoTMQTTPublishInfo_t * xMqttPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pxPublishInfo;
    az_iot_hub_client_twin_response xOutRequest;
    az_span xTopic = az_span_create( ( uint8_t * ) xMqttPublishInfo->pcTopicName,
                                     xMqttPublishInfo->usTopicNameLength );
    az_result xCoreResult;
    AzureIoTHubClientResult_t xReturn;
    uint32_t ulRequestID = 0;

    xCoreResult = 
        az_iot_hub_client_twin_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
                                                     xTopic, &xOutRequest );
    if( az_result_failed( xCoreResult ) )
    {
        AZLogWarn( ( "Twin topic parsing failed: 0x%08x", xCoreResult ) );
        xReturn = eAzureIoTHubClientFailed;
    }
    else
    {
        /* Verify the received publish is for the we have subscribed to. */
        AZLogInfo( ( "Twin topic: %.*s. with payload : %.*s",
                     mqttPublishInfo->topicNameLength,
                     mqttPublishInfo->pTopicName,
                     mqttPublishInfo->payloadLength,
                     mqttPublishInfo->pPayload ) );

        xReturn = eAzureIoTHubClientSuccess;

        if( pxContext->_internal.callbacks.xTwinCallback )
        {
            if( az_span_size( xOutRequest.request_id ) == 0 )
            {
                xMessage.xMessageType = eAzureIoTHubTwinDesiredPropertyMessage;
            }
            else if( az_result_failed( xCoreResult = az_span_atou32( xOutRequest.request_id, &ulRequestID ) ) )
            {
                if( ulRequestID & 0x01 )
                {
                    xMessage.xMessageType = eAzureIoTHubTwinReportedResponseMessage;
                }
                else
                {
                    xMessage.xMessageType = eAzureIoTHubTwinGetMessage;
                }
            }
            else
            {
                /* Failed to parse the message */
                AZLogError( ( "Request ID parsing failed: 0x%08x", xCoreResult ) );
                xReturn = eAzureIoTHubClientFailed;
            }

            if( xReturn )
            {
                xMessage.pvMessagePayload = xMqttPublishInfo->pvPayload;
                xMessage.ulPayloadLength = xMqttPublishInfo->xPayloadLength;
                xMessage.xMessageStatus = xOutRequest.status;
                xMessage.pucVersion = az_span_ptr( xOutRequest.version );
                xMessage.usVersionLength = ( size_t ) az_span_size( xOutRequest.version );
                xMessage.ulRequestID = ulRequestID;
                pxContext->_internal.callbacks.xTwinCallback( &xMessage,
                                                              pxContext->_internal.pvCallbackContext );
            }
        }
    }

    return xReturn;
}